Placeholder waveforms, sequence caching, segment filtering and further improvements in Pulsar

antsr opened the pull request:
Implements the use of placeholder waveforms for HDAWG8, that avoid writing large waveforms to csv files.

For long waveforms, this can lead to an order of magnitude reduction in upload times, see the plot below.
![20210124_195152_len_sweep_results](https://user-images.githubusercontent.com/1426628/105728722-c5671900-5f2c-11eb-8493-afd9899f79e7.png)

To control the use of the new feature, a boolean parameter `f'{AWG.name}_use_placeholder_waves'` is created.

Playback of analog waves tested by looking at output with an oscilloscope.
Playback of marker waves tested by looking at the waveforms on the web interface.

Example generated seqc code:
```
wave dfuvvbfapnkvst = placeholder(192);
assignWaveIndex(1, dfuvvbfapnkvst, 0);
wave dkgoibakcukzdl = placeholder(288);
assignWaveIndex(2, dkgoibakcukzdl, 1);
wave btbtluuevijuye = placeholder(384);
wave byduxxlwwxkuae = placeholder(384);
assignWaveIndex(btbtluuevijuye, byduxxlwwxkuae, 2);
wave imbzwppsajydu = placeholder(480, true);
assignWaveIndex(1, imbzwppsajydu, 3);
wave cnltljgpgkzfrp = placeholder(576, true);
assignWaveIndex(1, cnltljgpgkzfrp, 4);
wave pmifyojlialzr = placeholder(672, true);
wave dcgxspzxeyeihp = placeholder(672);
assignWaveIndex(pmifyojlialzr, dcgxspzxeyeihp, 5);
wave bkecqornotijfd = placeholder(768, true);
wave bfmbrfnhmalnvo = placeholder(768);
assignWaveIndex(bkecqornotijfd, bfmbrfnhmalnvo, 6);
wave dkeuttsztbbksb = placeholder(864, true);
assignWaveIndex(2, dkeuttsztbbksb, 7);
wave czeidcwdupapmz = placeholder(960);
wave bueqxwioynugjy = placeholder(960, true);
assignWaveIndex(czeidcwdupapmz, bueqxwioynugjy, 8);
wave dlcowakwkbyfva = placeholder(1056, true);
assignWaveIndex(2, dlcowakwkbyfva, 9);
wave qdblwwiqenwet = placeholder(1152);
wave caezlutwdhqcse = placeholder(1152, true);
assignWaveIndex(qdblwwiqenwet, caezlutwdhqcse, 10);
wave sakxswjqmkqon = placeholder(1248, true);
wave dsiijbiljtqrtm = placeholder(1248, true);
assignWaveIndex(sakxswjqmkqon, dsiijbiljtqrtm, 11);
wave btpzamyvpswtul = placeholder(1344, true);
wave bgcktcjqeytwor = placeholder(1344, true);
assignWaveIndex(btpzamyvpswtul, bgcktcjqeytwor, 12);
wave nkemynfhkhxys = placeholder(1440, true);
wave bkqaklruzamwew = placeholder(1440, true);
assignWaveIndex(nkemynfhkhxys, bkqaklruzamwew, 13);
wave bayygekgkycfgp = placeholder(1536, true);
wave brzmpsqvpbbrie = placeholder(1536, true);
assignWaveIndex(bayygekgkycfgp, brzmpsqvpbbrie, 14);



while (1) {
  // Segment seg0
  // Element default_ese_seg0
  waitDIOTrigger();
  // Segment seg1
  // Element default_ese_seg1
  waitDIOTrigger();
  playWave(1, dfuvvbfapnkvst);
  // Segment seg2
  // Element default_ese_seg2
  waitDIOTrigger();
  playWave(2, dkgoibakcukzdl);
  // Segment seg3
  // Element default_ese_seg3
  waitDIOTrigger();
  playWave(btbtluuevijuye, byduxxlwwxkuae);
  // Segment seg4
  // Element default_ese_seg4
  waitDIOTrigger();
  playWave(1, imbzwppsajydu);
  // Segment seg5
  // Element default_ese_seg5
  waitDIOTrigger();
  playWave(1, cnltljgpgkzfrp);
  // Segment seg6
  // Element default_ese_seg6
  waitDIOTrigger();
  playWave(pmifyojlialzr, dcgxspzxeyeihp);
  // Segment seg7
  // Element default_ese_seg7
  waitDIOTrigger();
  playWave(bkecqornotijfd, bfmbrfnhmalnvo);
  // Segment seg8
  // Element default_ese_seg8
  waitDIOTrigger();
  playWave(2, dkeuttsztbbksb);
  // Segment seg9
  // Element default_ese_seg9
  waitDIOTrigger();
  playWave(czeidcwdupapmz, bueqxwioynugjy);
  // Segment seg10
  // Element default_ese_seg10
  waitDIOTrigger();
  playWave(2, dlcowakwkbyfva);
  // Segment seg11
  // Element default_ese_seg11
  waitDIOTrigger();
  playWave(qdblwwiqenwet, caezlutwdhqcse);
  // Segment seg12
  // Element default_ese_seg12
  waitDIOTrigger();
  playWave(sakxswjqmkqon, dsiijbiljtqrtm);
  // Segment seg13
  // Element default_ese_seg13
  waitDIOTrigger();
  playWave(btpzamyvpswtul, bgcktcjqeytwor);
  // Segment seg14
  // Element default_ese_seg14
  waitDIOTrigger();
  playWave(nkemynfhkhxys, bkqaklruzamwew);
  // Segment seg15
  // Element default_ese_seg15
  waitDIOTrigger();
  playWave(bayygekgkycfgp, brzmpsqvpbbrie);
}
```

chellings committed:
sweep_functions: added MajorMinorSweep

chellings committed:
sweep_functions: added Transformed_Sweep

chellings committed:
swf.multi_sweep_function: automatically wrap qcodes parameters into sweep functions

chellings committed:
pulsar/segment/sequence: First draft of sweeping AWG nodes as hard sweep

antsr committed:
HDAWG8Pulsar: switch on unused AWG outputs with dummy sequence to enable DC offsets
i.e., the sequencer does not play anything on the output

antsr committed:
initial implementation of placeholder waveforms

antsr committed:
bugfix use_placeholder_waveforms
- tested working with analog waveforms
- marker waveforms not working yet

# Conflicts:
#	pycqed/measurement/waveform_control/pulsar.py

antsr committed:
bugfix placeholder waveforms marker channels

antsr committed:
again bugfix placeholder marker channels
working, tested by looking at waveforms on web interface

chellings committed:
first draft of segment filtering

antsr committed:
WIP: bugfixes and cleanup for segment filtering

chellings committed:
pulsar/sequence: first draft of sequence caching

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

chellings committed:
pulsar: improvements to sequence caching

chellings committed:
Segment.calculate_hash: added distortion dict and charge buildup compensation

chellings committed:
Pulsar: set userregs for segment filtering only in active AWGs

chellings committed:
segment: corrected hashables for sequence cache if reuse_waveforms is False

chellings committed:
bugfixes pulsar sequence cache

chellings committed:
pulsar: bugfixes in sequence chaching

chellings committed:
pulsar: additional logging for placeholder waveforms

chellings committed:
bugfix Segment.add_charge_compensation: make sure that element length is multiple of sample granularity

chellings committed:
Sequence: added attribute is_resolved avoid re-resolving in generate_waveforms_sequences

chellings committed:
Sequence: added method harmonize_element_lengths

chellings committed:
pulsar: clear sequence cache if compilation crashes

chellings committed:
QuantumExperiment: added argument harmonize_element_lengths

antsr committed:
pulsar placeholder waveforms: workaround to allow using only the first marker channel

chellings committed:
pulsar: clear sequence cache if a crash happens already during sequence resolution

chellings committed:
HDAWG8Pulsar placeholder waveforms: force reprogramming when pattern of reuse changes

antsr committed:
pulsar: bugfixes for placeholder waves.
This version was test successfully on the XLD setup.

antsr committed:
Pulsar:
  * change append_zeros to prepend_zeros
  + make prepend_zeros do so for only the first element of a segment

chellings committed:
Pulsar/UHFQC_Base: do not stop AWGs twice
This requires that the start() method of all other AWG accepts (and potentially ignores) a kwarg stop_first.

chellings committed:
QuantumExperiment: added filter_segments_mask to enable non-rectangular 2D sweep

chellings committed:
Pulsar: improved implementation of segement filtering with repeat pattern

chellings committed:
utilities.general: added get_pycqed_dir() to get the path of a pycqed app data dir

chellings committed:
Pulsar: reset sequence cache if another pulsar instance has programmed the AWGs

chellings committed:
pulsar: added trigger source ZSync for HDAWG and use validators to check for valid trigger source
Removed the ValueErrors in _zi_playback_string and _zi_interleaved_playback_string since the validators now take care of this.

chellings committed:
bugfix pulsar: reset sequence cache if prepend_zeros is changed

chellings committed:
bugfix AWG5014Pulsar: set channel_groups correctly (required by sequence caching code)

chellings committed:
HDAWG8Pulsar: new parameter {AWG}_prepend_zeros to overwrite prepend_zeros for an AWG or a sub-AWG
An int sets it for the AWG, a list of 4 ints sets it per sub-AWG, None keeps the global setting.

chellings committed:
HDAWG8Pulsar: use workaround for single channel playback also for second wave channel
(not only for standalone second marker channel)
This is required for compatibility with pulsar.{channel}_amplitude_scaling
because the HDAWG would otherwise play the wave2 signal via the first output path
while pulsar.{channel}_amplitude_scaling would scale the second output path.

chellings committed:
bugfix Pulsar sequence caching: detect previously active but now inactive channels on active AWGs

chellings committed:
Pulsar.program_awgs: auto-retry programming with empty sequence cache in case of an error

chellings committed:
bugfix QuantumExperiment: make harmonize_element_lengths work with sequence compression

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

chellings committed:
pulsar: fixed a merge bug (removed prefetching to work with current ZI firmware)

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

chellings committed:
Pulsar: made sequence_cache property private

chellings commented:
This branch meanwhile contains further features to speed up uploads and measurements:
- sequence caching in pulsar (to avoid re-programming AWGs if a sequence is the same as in the previous upload)
- segment filtering (to measure only a subset of the programmed segments)
- do not stop AWGs twice before starting a measurement

These additional features have been tested on XLD & BF1. Some features are not yet implemented to the Tektronix AWG (and some cannot be implemented for the Tektronix), but the existing features do not break anything when a Tektronix is part of the setup. Sequence caching works also for the Tektronix (but is effective only if all channels of the Tektronix remain unchanged).

The branch also includes:
- prepend_zeros (workaround for bugs in the HDAWG pipelined prefetching)
- ZSync trigger for the HDAWG

Moreover, this branch contains some experimental code for a feature to allow hard sweeping parameters that can be accessed via ZI nodes. This feature has so far been tested on a virtual setup, but it is hard to disentangle the changes from other changes in pulsar. Since tests on real setups have shown that this experimental code does not break anything in measurements that do not use it, we should merge it with this pull request to avoid future merge conflicts.

chellings commented:
Note: the branch now also contains the new version of the sweep_functions module. This includes the new sweep functions TransformedSweep and MajorMinorSweep, whic are not used in this branch, but also FilteredSweep, which is needed in this branch.

chellings committed:
bugfix Pulsar sequence caching: detect change of repeat pattern

chellings committed:
UHFQCPulsar/HDAWG8Pulsar segment filtering: cleaned up SeqC code

chellings committed:
bugfix UHFQC_multi_detector.finish: stop AWG
So far, only AWGs of the subdetectors were stopped, but UHFQC_multi_detector can have its own AWG.

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

chellings committed:
two_qubit_gates/single_qubit_gates: remove support for channels_to_upload
The new sequence caching feature of pulsar should be used instead.

chellings committed:
mqm/QuantumExperiment: removed use of channels_to_upload
The new sequence caching feature of pulsar should be used instead.

chellings committed:
SegmentSoftSweep: added deprecation warning for channels_to_upload
The new sequence caching feature of pulsar should be used instead.

chellings committed:
FilteredSweep and related code: comments and code cleanup

chellings committed:
MajorMinorSweep: improved docstring and comments

chellings committed:
Segment/Sequence: added comments for allow_filter and sweep_params
(sweep_params is an experimental feature)

chellings committed:
Pulsar: code cleanup and comments

chellings committed:
Pulsar: further cleanup and comments

chellings commented:
The pull request now also removes configuration of the channels_to_upload parameter from measurement functions/classes and adds a deprecation warning for this parameter. The new sequence caching feature of pulsar should be used instead.

I have pushed further commits with code clean-up. I will cherry-pick them onto Proj/S17 to allow us to run some final tests with the cleaned-up code before merging the PR. Running a single qubit calib plus the cell for a first fp scope should be enough to test all code branches that I touched in the clean-up. After that, I will approve the PR.

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

antsr commented on pycqed/measurement/measurement_control.py:
```diff
@@ -385,9 +394,25 @@ def measure_soft_adaptive(self, method=None):
+        """
+        :param filtered_sweep: (list of bools) indicates which of the
+            acquisition elements will be played by the AWGs (True) and which
+            ones will be skipped (False)
```
add what happens if `filtered_sweep=None` to the docstring

antsr commented on pycqed/measurement/measurement_control.py:
```diff
@@ -385,9 +394,25 @@ def measure_soft_adaptive(self, method=None):
         new_data = np.array(self.detector_function.get_values()).T
 
+        if filtered_sweep is not None:
+            # Fill the data points that were not measured with NaN.
```
This confused me for a bit, as "fill" implied that there already is a slot for the missing data in the array.
In fact, we are adding the missing data here, increasing the array size.

antsr commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -75,6 +77,9 @@ def __init__(self, dev=None, qubits=None, operation_dict=None,
             sequence_kwargs (dict): keyword arguments passed to the sequence_function.
                 see self._prepare_sequences()
+            filter_segments_mask (array of bool): An array with dimension
+                n_0 x n_1, where n_i is the number of sweep points in
```
does this include calibration points?

chellings replied:
I have added further details to the docstring.

antsr commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -95,6 +100,9 @@ def __init__(self, dev=None, qubits=None, operation_dict=None,
                 for the second dimension.
+            harmonize_element_lengths (bool, default False): whether it
+                should be ensured for all AWGs and all elements that the
+                element length is the same in all sequences.
```
Adding the typical use case here would be useful for the user to know the context (to avoid sequence update)

antsr commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -501,6 +510,9 @@ def _configure_mc(self, MC=None):
                                 "for now.")
 
+        if self.harmonize_element_lengths:
+            self.sequences[0].harmonize_element_lengths(self.sequences)
```
You should call the static method from the class, not the instance.

chellings replied:
This technique is also used at other points in pycqed, e.g. for merging sequences. The point is that the staticmethod is available wherever the sequence object is available, without the need to explicitly include the module. Moreover, we could in the future decide to have special kinds of sequences which bring their own overridden staticmethods if needed. In this case, calling the staticmethod via the instance automatically gives the right version of the staticmethod. If you have strong arguments against this way of implementing it, we should discuss it also for other points in pycqed where the same is done. Otherwise, I would just keep it as it is now.

antsr replied:
I still think that the current code is confusing to read, as it assigns a special role to the first sequence `sequences[0]`.
Imports are handeled automatically by most IDEs, so I don't really see the large downside of having an extra import for the class.
We could look at the other instances where this is done as well.

antsr commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -520,19 +532,48 @@ def _configure_mc(self, MC=None):
+                        # measure nothing (by setting last < first)
+                        lookup[sp] = (1, 0)
+                sweep_func_2nd_dim = swf.FilteredSweep(
+                        self.sequences[0], lookup, [sweep_func_2nd_dim])
```
Adding a few comments about what is the goal of each block here could be useful.
The way I understand it (maybe that's wrong) there are two main steps:
 * First, you exclude some of the segments (from the end of the sequence) from the filter. This resulting sequence code will be different from having `True` in the mask. Would it make sense to have a more flexible way to do this? For example have `None` in the filter mask or something like that.
 * Then you create a "lookup table", which is a map from soft sweep point value? to a range of hardware sweep points to measure. I think naming this variable something more clear, like "filter_ranges" would be more clear. Also, its unclear to me, why you index the map by the value of the mc_point, not the index. 

Is it currently true that if I have a filter [False, True, False, True, False], it will measure the 3 segments in the middle and then fail, because it got a wrong number of data points?

chellings replied:
Yes, it will measure the three points in the middle, but no, it will not fail. You can think of this as a way of automatically correcting the mask to something that is supported by the lower layers. I have added further comments to the code and I have renamed the variable to the same name as in FilteredSweep.

antsr commented on pycqed/measurement/sweep_functions.py:
```diff
@@ -303,3 +303,107 @@ def __init__(self, sweep_function, values, name=None, parameter_name=None,
+                 major_values,
+                 name=None,
+                 parameter_name=None,
+                 unit=None):
+        super().__init__()
+
+        self.major_sweep_function = \
+            mc_parameter_wrapper.wrap_par_to_swf(major_sweep_function) \
+                if isinstance(major_sweep_function, qcodes.Parameter) \
+                else major_sweep_function
+        self.minor_sweep_function = \
+            mc_parameter_wrapper.wrap_par_to_swf(minor_sweep_function) \
+                if isinstance(minor_sweep_function, qcodes.Parameter) \
+                else minor_sweep_function
+        if self.major_sweep_function.sweep_control != 'soft' or \
+                self.minor_sweep_function.sweep_control != 'soft':
+            raise ValueError('Offset_Sweep: Only software sweeps supported')
+        self.sweep_control = 'soft'
+        self.major_values = np.array(major_values)
+        self.parameter_name = self.major_sweep_function.parameter_name \
+            if parameter_name is None else parameter_name
+        self.name = self.major_sweep_function.name if name is None else name
+        self.unit = self.major_sweep_function.unit if unit is None else unit
+
+    def prepare(self, *args, **kwargs):
+        self.major_sweep_function.prepare(*args, **kwargs)
+        self.minor_sweep_function.prepare(*args, **kwargs)
+
+    def finish(self, *args, **kwargs):
+        self.major_sweep_function.finish(*args, **kwargs)
+        self.minor_sweep_function.finish(*args, **kwargs)
+
+    def set_parameter(self, val):
+        # find the closes allowed value of the major_sweep_function
+        ind = np.argmin(np.abs(self.major_values - val))
+        mval = self.major_values[ind]
+        self.major_sweep_function.set_parameter(mval)
+        # use the minor_sweep_function to bridge the difference to the
+        # target value
+        self.minor_sweep_function.set_parameter(val - mval)
```
Not sure if this class is really relevant for this PR, but it looks ok to me.

antsr commented on pycqed/measurement/sweep_functions.py:
```diff
@@ -303,3 +303,107 @@ def __init__(self, sweep_function, values, name=None, parameter_name=None,
+
+    def set_parameter(self, val):
+        # Determine the current segment filter and inform Pulsar.
+        filter_segments = self.filter_lookup[val]
```
I see now why the lookup table had to index by value

antsr commented on pycqed/measurement/sweep_functions.py:
```diff
@@ -303,3 +303,107 @@ def __init__(self, sweep_function, values, name=None, parameter_name=None,
+        self.minor_sweep_function.set_parameter(val - mval)
+
+
+class FilteredSweep(multi_sweep_function):
```
Does it make sense to inherit from `multi_sweep_function`? Do we ever use the functionality that `multi_sweep_function` implements? That is, setting multiple sweep functions to the same value.

chellings replied:
I wanted to give the user the possibility to directly use a filtered sweep in combination with multiple sweep functions without adding a extra hierarchy level (to keep the hierarchy of sweep functions flat, i.e., to avoid thing like MC.sweep_functions[1].sweep_function.sweep_functions[0] etc).

The code in QE currently does not make use of this: for the sake a unified treatment of all kinds of soft sweep functions, I accepted the drawback of adding another hierarchy layer in the code in QE). However, I think it does not hurt to give the user the possibility to use the FilteredSweep directly as a multi sweep function. Or do you think it has significant drawbacks?

antsr replied:
I think my comment was more about `multi_sweep_function` in general -- the functionality of `multi_sweep_function` seems to be that it sets the same sweeppoint value to multiple sweep functions. For parallel sweeps, functionality to set different values to different sweep functions seems much more useful.

chellings replied:
Yes, I use multi sweep function a lot in the ParallelLOSweepExperiment code on the S17 branch. Will be one of the next PRs, but I cannot create that PR yet due to conflicts with open PRs.

antsr commented on pycqed/measurement/sweep_functions.py:
```diff
@@ -303,3 +303,107 @@ def __init__(self, sweep_function, values, name=None, parameter_name=None,
+        # The filtered_sweep property stores a mask indicating which
+        # acquisition elements are recorded (will be accessed by MC to
+        # handle the acquired data correctly).
+        seg_mask = np.invert(self.allow_filter)
```
I guess functionally you are interested in a logical not here, not bitwise not.

antsr commented on pycqed/measurement/waveform_control/segment.py:
```diff
@@ -46,6 +46,18 @@ def __init__(self, name, pulse_pars_list=[]):
         self.distortion_dicts = {}
+        # The sweep_params dict is processed by generate_waveforms_sequences
+        # and allows to sweep values of nodes of ZI HDAWGs in a hard sweep.
+        # Keys are of the form awgname_chY_nodename (with _ instead of / in
```
Here it is unclear what chY refers to -- is it the channel name or channel id? Do we not separate between the two anymore?

chellings replied:
It is the chid. I have changed the comment accordingly.

antsr commented on pycqed/utilities/general.py:
```diff
@@ -800,3 +800,14 @@ def find_symmetry_index(data):
         corr.append((data_filtered*data_filtered[::-1]).sum())
     return np.argmax(corr), corr
+
+def get_pycqed_dir():
```
from the function name I would guess that this function returns the path of the repository

chellings replied:
I have changed it to get_pycqed_appdata_dir.

antsr commented on pycqed/measurement/waveform_control/segment.py:
```diff
@@ -1098,24 +1110,36 @@ def get_element_channels(self, element, awg=None):
         else:
             hashlist.append(self.pulsar.clock(channel=channel))  # clock rate
             for par in ['type', 'amp', 'internal_modulation']:
                 try:
                     hashlist.append(self.pulsar.get(f'{channel}_{par}'))
                 except KeyError:
                     hashlist.append(False)
```
Not part of this PR, but could be written as `hashlist.append(self.pulsar.get(f'{channel}_{par}', False))`

chellings replied:
Sure, you can push this improvement to the PR branch.

antsr commented on pycqed/measurement/waveform_control/sequence.py:
```diff
@@ -53,19 +54,38 @@ def extend(self, segments):
+                channel
+        :param awgs: a list of AWG names. If None, waveforms will be
+            generated for all AWGs used in the sequence.
+        :param get_channel_hashes: (bool, default: False) do not create
+            waveforms, and instead return a dict of channel-elements indexed
+            by channel names, each channel-element consisting of a
+            waveform-hash for each codeword on that channel.
```
Is this structure the same as a flattened `sequences`? A comment on how this differs from `sequences` could be useful.

chellings replied:
I have added further info to the docstring. It is not flattened, but
sequences[awg][elname][cw][chid] == channel_hashes[ch][elname][cw]

antsr commented on pycqed/measurement/waveform_control/sequence.py:
```diff
@@ -77,26 +97,99 @@ def generate_waveforms_sequences(self, awgs=None):
-                                                 .popitem()[1].popitem()[1]
+                            if get_channel_hashes:
+                                if ch not in channel_hashes:
+                                    channel_hashes[ch] = {}
```
Do you rely on the fact that python 3.6+ maintains insertion order in dicts here?
If yes, then I would find it a bit more clear to use an `odict`.

chellings replied:
Sorry, but I do not see which part of the code makes an assumption about insertion order. This line of code just ensures that I do not add to a sub-dict that does not exist yet. Can you please explain the issue that you spotted in more detail?

antsr replied:
I can't see what I had in mind with this comment anymore.

antsr commented on pycqed/measurement/waveform_control/sequence.py:
```diff
@@ -77,26 +97,99 @@ def generate_waveforms_sequences(self, awgs=None):
                         sequences[awg][elname]['metadata']['acq'] = False
-        return waveforms, sequences
-                
+                    if seg.allow_filter is not None:
```
Maybe I missed it, but in what cases is seg.allow_filter `None`? It defaults to `False`.

chellings replied:
Probably an artifact from an older version of the code. I have removed the check.

antsr commented on pycqed/measurement/waveform_control/sequence.py:
```diff
@@ -77,26 +97,99 @@ def generate_waveforms_sequences(self, awgs=None):
+        elements that the element length is the same in all sequences. This is
+        done by setting the length of each AWG element to the maximum length
+        across all sequences. After this, overlap is checked and the sequences
+        are markes as resolved.
```
marked

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -33,17 +35,25 @@
     class
     """
     _supportedAWGtypes = (UHFQC, dummy_UHFQC)
+    _num_awgs = 1
```
unused?

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -130,11 +140,16 @@ def _create_awg_parameters(self, awg, channel_name_map):
+            group.append(name)
+        # all channels are considered as a single group
+        for name in group:
+            self.channel_groups.update({name: group})
```
`self.channel_groups[name] = group`

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -221,54 +238,122 @@ def _program_awg(self, obj, awg_sequence, waveforms, repeat_pattern=None):
+            playback_strings += self._zi_playback_string(
+                name=obj.name, device='uhf', wave=wave, acq=acq,
+                allow_filter=(
+                        allow_filter and metadata.get('allow_filter', False)))
```
why is the `and` needed here?

chellings replied:
To use segment filtering, both the kwarg of play_element and the metadata of the element have to allow filtering. I have added a comment to explain this.

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -221,54 +238,122 @@ def _program_awg(self, obj, awg_sequence, waveforms, repeat_pattern=None):
+                        allow_filter[seg_indices[-1]] += 1
+                else:  # segment
+                    seg_indices.append(index)
+                    allow_filter[seg_indices[-1]] = 0
```
`allow_filter[index] = 0` maybe?

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -221,54 +238,122 @@ def _program_awg(self, obj, awg_sequence, waveforms, repeat_pattern=None):
+                # repetitions via a user register to the SeqC variable last_seg
+                repeat_pattern = ('last_seg', 1)
+
+            def repeat_func(n, el_played, index, playback_strings,
```
I don't understand what this helper function does anymore. A brief comment would be useful.

chellings replied:
I have now added comments that reflect what I understand from reverse-engineering the original function. Feel free to improve the comments if they are not accurate enough.

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -632,43 +790,146 @@ def _program_awg(self, obj, awg_sequence, waveforms, repeat_pattern=None):
-            # class in 
+            # tell ZI_base_instrument.start() to start this sub AWG
             obj._awg_needs_configuration[awg_nr] = False
             obj._awg_program[awg_nr] = True
```
The meaning of these flags could be explained a bit.

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -587,38 +705,78 @@ def _program_awg(self, obj, awg_sequence, waveforms, repeat_pattern=None):
-                                    for ch in [ch1id, ch1mid, ch2id, ch2mid])
-                        wave_definitions += self._zi_wave_definition(wave,
-                                                                defined_waves)
+                        wave_idx_lookup[element][cw] = {}
```
this line looks like its not needed

antsr replied:
Maybe it should default to `None` if we don't use placeholder waves?

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -632,43 +790,146 @@ def _program_awg(self, obj, awg_sequence, waveforms, repeat_pattern=None):
+                    run_compiler = True
+
+            if run_compiler:
+                # We have to retrieve the folllowing parameter to set it
```
following

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -1073,8 +1357,15 @@ def __init__(self, name='Pulsar', master_awg=None):
+                           set_parser=self._use_sequence_cache_parser)
+        self.add_parameter('prepend_zeros', initial_value=0, vals=vals.Ints(),
                            parameter_class=ManualParameter)
+        # keep old parameter name for backwards compatibility
+        self.add_parameter('append_zeros', initial_value=0, vals=vals.Ints(),
+                           set_cmd=(lambda v, self=self: self.prepend_zeros(v)),
+                           get_cmd=(lambda self=self: self.prepend_zeros()))
```
Maybe we can remove the old parameter now? I guess prepending worked better in combination with segment filter.

chellings replied:
Yes, you could push this to the PR branch.

chellings replied:
Note: There will be another PR concerning pulsar soon, and we could also do it as part of this other PR. It would be good if we could move on with #210 ASAP @antsr

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -1083,18 +1374,25 @@ def __init__(self, name='Pulsar', master_awg=None):
                            initial_value=None, parameter_class=ManualParameter)
         self.add_parameter('flux_crosstalk_cancellation_shift_mtx',
                            initial_value=None, parameter_class=ManualParameter)
+        self.add_parameter('filter_segments',
```
Please add a brief comment

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -1104,6 +1402,38 @@ def __init__(self, name='Pulsar', master_awg=None):
+
+    def reset_sequence_cache(self):
+        self._sequence_cache = {}
+        self._sequence_cache['settings'] = {}
+        self._sequence_cache['metadata'] = {}
+        self._sequence_cache['hashes'] = {}
+        self._sequence_cache['length'] = {}
```
A brief description of these dictionaries somewhere would be very useful.

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -1104,6 +1402,38 @@ def __init__(self, name='Pulsar', master_awg=None):
+        self._sequence_cache['hashes'] = {}
+        self._sequence_cache['length'] = {}
+
+    def check_for_other_pulsar(self):
```
Ideally we would also want to detect manual change of the sequences. Could the id be stored in the AWG sequence code?

chellings replied:
We could consider adding this in the future. You could create an issue to make sure that we'll remember it.

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -1275,6 +1613,15 @@ def stop(self):
             self._stop_awg(awg)
     
     def program_awgs(self, sequence, awgs='all'):
+        try:
+            self._program_awgs(sequence, awgs)
+        except Exception as e:
+            log.warning(f'Pulsar: Exception {repr(e)} while programming AWGs. '
+                        f'Retrying after resetting the sequence cache.')
+            self.reset_sequence_cache()
+            self._program_awgs(sequence, awgs)
```
The try-except doesn't really hurt here, but is this still needed?

chellings replied:
Since sequence caching is am optional speedup feature, it makes sense to fall back to the safer version of not using the cache if something goes wrong. This retry is relatively cheap and might prevent crashed measurements, so I would keep it.

antsr commented on pycqed/measurement/waveform_control/pulsar.py:
```diff
@@ -1291,33 +1638,199 @@ def program_awgs(self, sequence, awgs='all'):
+            # channels in the same group) will be re-programmed, while other
+            # channels can be re-uploaded by replacing the existing waveforms.
+            ch_length = {}
+            for ch, hashes in channel_hashes.items():
+                ch_awg = self.get(f'{ch}_awg')
+                if ch_awg in awgs_to_program + awgs_with_channels_to_upload:
+                    ch_length[ch] = {
+                        elname: {cw: len(waveforms[h]) for cw, h in el.items()}
+                        for elname, el in hashes.items()}
+                # Checking whether programming is done only for channels that
+                # are marked to be uploaded but not yet marked to be programmed
+                if ch not in channels_to_upload or ch in channels_to_program \
+                        or ch_awg in awgs_to_program:
+                    continue
+                try:
+                    np.testing.assert_equal(
+                        sequence_cache['length'].get(ch, {}),
+                        ch_length[ch])
+                except AssertionError:  # changed length or sequence structure
+                    for c in self.channel_groups[ch]:
+                        channels_to_program.append(c)
+            # update the length cache
+            for ch in channels_to_program:
+                sequence_cache['length'][ch] = ch_length.get(ch, {})
+            # Update the cache for channels that are on an AWG marked for
+            # complete re-programming (these channels might have been skipped
+            # above).
+            for ch in self.channels:
+                if self.get(f'{ch}_awg') in awgs_to_program:
+                    sequence_cache['settings'][ch] = settings.get(ch, {})
+                    sequence_cache['hashes'][ch] = channel_hashes.get(
+                        ch, {})
+                    sequence_cache['metadata'][ch] = metadata.get(ch, {})
+                    sequence_cache['length'][ch] = ch_length.get(ch, {})
+            log.debug(f'awgs_to_program = {repr(awgs_to_program)}\n'
+                      f'awgs_with_channels_to_upload = '
+                      f'{repr(awgs_with_channels_to_upload)}\n'
+                      f'channels_to_upload = {repr(channels_to_upload)}\n'
+                      f'channels_to_program = {repr(channels_to_program)}'
+                      )
```
This part of the code is too complicated for me to review functionality-wise, which should be still ok as we have tested it a lot. Formatting- / structure-wise it looks ok.

antsr reviewed (commented):
I can only comment as the creator of the PR. I think a few small things (see comments) could be improved to make the code easier to handle in the future.

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms

chellings committed:
MC.measure_hard: improved comments/docstring

chellings committed:
bugfix FilteredSweep: use logical not instead of bitwise not

chellings committed:
Pulsar: code cleanup and improved comments

chellings committed:
utilities.general: renamed get_pycqed_dir to get_pycqed_appdata_dir

chellings committed:
Sequence/Segment: code cleanup and improved comments

chellings committed:
QuantumExperiment: improved comments and renamed a variable

chellings committed:
UHFQCPulsar: improved comments in the code for resolving repeat patterns

chellings reviewed (commented):
@antsr Please see the new commits that I have pushed. I have marked conversations as resolved in cases where my committed changes directly resolve your comments. For all other comments, please see my responses.

antsr committed:
minor refactors

antsr committed:
bugfix PR210

antsr commented:
Tested on XLD1 (single qubit calib and FPscope with segment filtering are working), and ready to merge from my side.

chellings commented on pycqed/utilities/get_default_datadir.py:
```diff
@@ -2,26 +2,18 @@
-            datadir))
+    # If the mac_address is unknown
+    # Stores data in the default data location (pycqed_py3/data/)
+    datadir = os.path.abspath(os.path.join(
```
Note that this is actually a bugfix for the already merged PR #180. Note for the future: always double-check to not break any imports when deleting modules. @grahamnorris @antsr 

Please also change the comment and docstring to reflect the new behavior of the function @antsr
Thank you!

antsr committed:
get_default_datadir: format docstring, remove print statement

antsr committed:
Segment.calculate_hash: bugfix hashlist fix

chellings committed:
Merge branch 'qudev_master' into devel/placeholder_waveforms
